Sqs and perception (#1041)
o3de/o3de.org: content/docs/user-guide/gems/reference/kythera-ai/spatial-query-system.md

        

## Writing spatial queries

Each level in a game can include a file called `SpatialQueries.sqs` which defines a collection of Spatial Queries. Queries common to multiple levels can be placed in the global Scripts folder. Whenever a level is loaded, all queries are first unloaded and then the global queries are loaded in. The level-specific queries come next and will overwrite any global queries with the same name.

There is a console command, `kyt_ReloadSpatialQueries`, which will reload the queries in the file on-demand, allowing rapid iteration during development.
In addition to the queries written in C++ as shown above, queries can also be provided in a file. These files will be parsed by the Spatial Query System, and used to construct spatial queries. The syntax for these queries is the same as for C++ queries.

### Query structure

A spatial query consists of four blocks - `Generators`, `Conditions`, `Scores` and `Fallbacks`. When the query is executed:

> *   The Generators are run in parallel (via time-slicing) and each returns one or more candidate positions or entities (such as cover points) as output

Reviewer: If this is a guaranteed ordering of steps, would recommend using a numbered list.

> *   The Generators are run in parallel (via time-slicing) and each returns one or more candidate positions or entities (such as cover points) as output
>     
> *   Once all Generators have completed, this set of candidate positions is passed through each Condition, in order, and any that fail are immediately dropped from further processing

Reviewer: ```suggestion
*   The Generators are run in parallel (via time-slicing) and each returns one or more candidate positions or entities (such as cover points) as output.
    
*   Once all Generators have completed, this set of candidate positions is passed through each Condition, in order, and any that fail are immediately dropped from further processing.
```

> *   Each remaining candidate position is now assigned a score by evaluating each Score criterion. These consist of a weight and a criterion that produces a score. Each criterion score is multiplied by its weight; the sum of all these values is the total score assigned to the position.

Reviewer: ```suggestion
*   Each remaining candidate position is assigned a score by evaluating each individual Score criterion. These consist of a weight and a criterion that produces a score. Each criterion score is multiplied by its weight; the sum of all these values is the total score assigned to the position.
```

Not sure what's up with the stray caps of `Score` here - could be a typo or an un-bolded term?

> *   The candidate position with the highest score wins and is returned as the chosen position.
>     
> *   If no position passed all of the conditions, the query will use the Fallbacks in the order they are provided.

Reviewer: ```suggestion
*   The candidate position with the highest score wins and is returned as the chosen position. If no position passed all of the conditions, the query will use the Fallbacks in the order they are provided.
```

    

### Query syntax definition

> Spatial query syntax has been designed to be intuitive and legible. The syntax of the query definition itself is given below in Backus–Naur Form.

Reviewer: If this follows IETF RFC 5234 (the only freely available formalization of BNF syntax I could find):

```suggestion
Spatial query syntax has been designed to be intuitive and legible. The syntax of the query definition itself is given below in [Backus–Naur Form](https://datatracker.ietf.org/doc/html/rfc5234).
```

<generator-args>    ::= <args-list> | <named-args-list>

<conditions-block>  ::= "Conditions {" <condition-list> "}"
<condition-list>    ::= <condition> | <condition> "," <condition-list>
<condition>         ::= <function-name> "(" <condition-args> ")"
<condition-args>    ::= <args-list> | <condition-list>

<scores-block>      ::= "Scores {" <score-list> "}"
<score-list>        ::= <score> | <score> "," <score-list>
<score>             ::= "Weight(" <weight-multiplier> "," <condition> ")"
<weight-multiplier> ::= <float>

<args-list>         ::= <arg> | <arg> "," <args-list>
<named-args-list>   ::= <named-arg> | <named-arg> "," <named-args-list>
<named-arg>         ::= <argument-name> "(" <arg> ")"

Unspecified symbols:
  <arg>             Any C++ value that is a valid argument to the function
  <argument-name>   Any C++ symbol that the generator recognizes
  <float>           A C++ floating point number
```

### General concepts

*   `SpatialQueryConstruction` namespace
    
    *   The types referred to in this document are all declared within this namespace, so you'll either need a `using` declaration or to scope them appropriately.
        
*    Objects
    

>     *   In several places, a query requires you to specify an object. This is used for things like the object around which to generate points, doing a raycast from and to specific objects, calculating distance to some object and so on. Possible values are:

Reviewer: ```suggestion
    *   In several places, a query requires you to specify an object. This object is used for things like point generation, raycasting, and calculating distance.  Possible values are:
```

        
        *   Querier: Refers to the agent running the query and is the default value.
            
        *   Target: Refers to the querier's current target entity. 
            
        *   Reference: If a reference object (entity or position) is provided to the query, this allows you to use it in a function. This is useful for taking additional information into account, such as finding cover near some reference point while avoiding line-of-sight to the target.
            
        *   Point: Refers to the current point under consideration while calculating conditions and weights, and is not valid in a generator.
            
*   Context
    

>     *   A query has a general context object that is available to criteria as they execute. This is used to track progress through the query execution and provides access to general data such as the agent that requested the query.

Reviewer: ```suggestion
    *   A query has a general context object that is available to criteria as they execute. The query context tracks progress through the query execution and provides access to general data such as the agent that requested the query.
```

>     *   Each candidate position also has a context of its own, which is made directly available to the criteria as they process the position. The context tracks progress through condition and weight expressions and can be used to store data between updates.

Reviewer: ```suggestion
    *   Each candidate position also has a context of its own, which is made directly available to the criteria as they process the position. Object contexts tracks progress through condition and weight expressions and can be used to store data between updates.
```

            
        *   AvoidDynamicObstacles: bool (default false)
            
        *   OpenAreasOnly: bool (default false) True can speed up results but skips points close to unnavigable regions
            
        *   PointRadius: float (default 0.0) Causes points too close to each other to be removed
            

* * *

### Logical functions  

#### Direct logical operations

*   `LessThan(lhs, rhs)`
    
*   `LessThanOrEqualTo(lhs, rhs)`
    
*   `GreaterThan(lhs, rhs)`
    
*   `GreaterThanOrEqualTo(lhs, rhs)`
    
*   `Equal(lhs, rhs)`
    
*   `NotEqual(lhs, rhs)`
    
*   `Not(boolean)`
    

#### Grouped logical operations

*   `AllOf(expressions...)`
    
    *   Returns true if every expression is true. Will return early if any expression returns false.
        
*   `AnyOf(expressions...)`
    
    *   Returns true if any single expression returns true. Will return early if any expression returns true.
        
*   `NoneOf(expressions...)`
    
    *   Returns true if none of the expressions return true. Will return early if any expression returns true.
        

* * *

> ### Operation functions

Reviewer 2: I'm not sure if the bold formatting was meant to double as a grouping for categorization purposes, but if it was, it's not going to be obvious to everyone. Therefore I've replaced all the bold with code formatting. If you want to categorize, I recommend explicitly inserting category labels. If you don't want to categorize, then I'd recommend alphabetizing the whole list.

#### **Weight(multiplier, function)**

*    The weight multiplier parameter must be either a float or a boolean. 
    

#### `Chance(probability)`

*   0 <= `probability` <= 1
    
    *   Returns true if a generated random number in the range of 0-1 is below the given probability value.
        

#### `HasTag(tag)`

*   Returns true if the current point has the given tag.
    

#### `Distance3D(object)`

*   Returns the 3D distance from a point to an object. 
    

#### **DistanceClosestWithRelationship(maxDistance, ESQSRelationship)**

*   Returns the 3D distance from a point to the closest entity with a specified relationship with Querier. 
    

#### `DistanceClosestFriendly(maxDistance)`

*   Returns the 3D distance from a point to the closest entity with a Friendly relationship with Querier. 
    

#### `DistanceClosestHostile(maxDistance)`

*   Returns the 3D distance from a point to the closest entity with a Hostile relationship with Querier. 
    

> #### **IsWithinRangeWithRelationship(range, ESQSRelationship)**

Reviewer 2: ```suggestion
#### `IsWithinRangeWithRelationship(range, ESQSRelationship)`
```

*   Returns true if there is any entity within range with the given relationship.
    

#### `IsFriendlyWithinRange(range)`

*   Returns true if there is any friendly entity within range.
    

#### `IsHostileWithinRange(range)`

*   Returns true if there is any hostile entity within range.
    

> #### **Raycast(object, desiredResult, sourceOffset, objectOffset)**

Reviewer 2: ```suggestion
#### `Raycast(object, desiredResult, sourceOffset, objectOffset)`
```

*   Performs physics raycast from point to object, with configurable offsets for both the start and endpoint.
    

#### `CanShootTargetHead`

*   Specialization of raycast. Performs raycast from point with offset at Querier eye level, to Target's eye level. Returns true if no hits.
    

#### `CanShootTargetTorso`

*   Specialization of raycast. Performs raycast from point with offset at Querier eye level, to Target's torso level. Returns true if no hits.
    

#### `EnemyLineOfSightBlocked`

*   Specialization of raycast. Performs raycast from point with offset at Querier's torso level, to Target's eye level. Returns true if any hits.
    

> #### **IsCurrentlyOwnedBy(slotLabel, object)**

Reviewer 2: ```suggestion
#### `IsCurrentlyOwnedBy(slotLabel, object)`
```

*   Returns true if the current cover point is owned by the given object.
    

#### `CanReachBeforeTarget`

*   Returns true if the distance from a point to Querier is less than the distance from the point to Target.
    

#### `EyeOffset(object)`

*   Returns the offset of the object's eye as a KytVec3.
    

#### `TorsoOffset(object)`

*   Returns the offset of the object's torso as a KytVec3.
    

> #### **RelativeDirection(object, ERelativeDirection)**

Reviewer 2: ```suggestion
#### `RelativeDirection(object, ERelativeDirection)`
```

*   Returns the relative direction (forward, backward, left, right, up, or down) of the object as a KytVec3.
    

#### `DirectionTo(object)`

*   Get a direction vector from the object to the candidate position.
    

#### `DirectionFrom(object)`

*   Get a direction vector from the candidate position to the object.
    

> #### **DotProduct(vec1, vec2)**

Reviewer 2: ```suggestion
#### `DotProduct(vec1, vec2)`
```

*   Returns the dot product of the two vectors.
    

#### `Directness`

*   Returns the directness of a given point: (`distanceFromQuerierToTarget` - `distanceFromPointToTarget`) / `distanceFromQuerierToPoint`)
    

> #### **NormaliseAndClamp(function, min, max)**

Reviewer 2: ```suggestion
#### `NormaliseAndClamp(function, min, max)`
```

*   Normalises and clamps the float given by Function within the range of \[Min, Max\].
    
    *   Named parameters:
        
        *   function: float
            
        *   min: float
            
        *   max: float
            

#### `Absolute(function)`

*   Returns the absolute value of the result from the input function. 
    

#### `Heading`

*   Returns the cosine of the angle between the querier's forward direction and the direction to the current point.
    

#### `IsCoverOfType(coverType)`

*   Returns true if the current point has the given cover type.
    

> #### **NavRaycast(object, desiredResult)**

Reviewer 2: ```suggestion
#### `NavRaycast(object, desiredResult)`
```

*   Performs navmesh raycast from point to object. 
    

> #### **OctreeRaycast(Object, DesiredResult, ShipRadius)**

Reviewer 2: ```suggestion
#### `OctreeRaycast(Object, DesiredResult, ShipRadius)`
```

*   Performs octree raycast from point to Object.
    

> #### **NavDistance(object, maxPathDistance)**

Reviewer 2: ```suggestion
#### `NavDistance(object, maxPathDistance)`
```

*   Calculates the pathing distance between the current point and the given object. 
    

#### `OctreeSurfaceDistance(Object)`

*   Calculates the approximate distance between the current point and the nearest octree surface
    
    *   Octree parameter "MaxRadius" needs to be high for this to be useful some distance from octree surfaces
        
    *   Object parameter should be Querier (it is only used to obtain the octree)
        

> #### **GetBlackboard\[\*\]**

Reviewer 2: ```suggestion
#### `GetBlackboard\[\*\]`
```

*   Returns a value of specified type dependant on the function listed below, from the entity blackboard under the specified key.
    
    *   Named Parameters: 
        
        *    `Key (StringHash)`
            
    *   Available function types:
        
        *   _**GetBlackboardFloat**_(`Key`)
            
        *   _**GetBlackboardInt**(_`Key`)
            
        *   _**GetBlackboardBool**_(`Key`)
            
        *   _**GetBlackboardStringHash**_(`Key`)
            
        *   _**GetBlackboardPos**_(`Key`)
            
        *   _**GetBlackboardVec3**_(`Key`)

## Debugging spatial queries

### Debugging parser errors

> When working with parsed Spatial Queries (either from a .sqs file in Lumberyard, or an SQS Asset in Unreal Engine), you may sometimes make errors when writing your queries. This can lead to errors that cause your spatial query to be unparseable, meaning that Kythera will not be able to use it.

Reviewer 2: Since this was written for O3DE, references to other engines are not necessary here.
```suggestion
When working with parsed Spatial Queries, you may sometimes make errors when writing your queries. This can lead to errors that cause your spatial query to be unparseable, meaning that Kythera will not be able to use it.
```

In these situations, we output error logging to the Console:

![SQS errors logged to the console](/images/user-guide/gems/kythera-ai/sqs-console.png)

If you have modified your spatial query and it no longer runs, be sure to check here first.

### Visual debugging

Visual debugging of the spatial query system is available in Kythera. This is enabled on a per-entity basis from within the Kythera Inspector, by setting **Debugging**→**DebugDraw**→**SpatialQuery**→**Enable** to true for that entity:

> ![Visual debugging of spatial queries with the Kythera AI Inspector](/images/user-guide/gems/kythera-ai/sqs-visual-debugging)

Reviewer 2: ```suggestion
![Visual debugging of spatial queries with the Kythera AI Inspector](/images/user-guide/gems/kythera-ai/sqs-visual-debugging.png)
```

> (Note that global debug draw must also be enabled - either by using the **Enable Debug Draw** button in the Lumberyard Editor, or by setting console variable `kyt_DrawMaster` to 1.)
> 
> Once this is enabled, debug visualization will be drawn into your scene. 

Reviewer 2: Use note shortcode.
```suggestion
{{< note >}}
Global debug draw must also be enabled - either by using the **Enable Debug Draw** button in the O3DE Editor, or by setting console variable `kyt_DrawMaster` to 1. Once this is enabled, debug visualization will be drawn into your scene. 
{{< /note >}}
```

> ![Debug draw of spatial queries](/images/user-guide/gems/kythera-ai/sqs-debug-visualization.png)

Reviewer 2: Rename the png included with this PR to match this link (`sqs-debug-visualization.png`).

*   Points which fail a Condition are drawn as gray circles, and labelled with the name of the condition criterion that they failed.
    
    *   Note that the name printed is the name of the outermost function in a Condition block. Often you will see lots of points labelled with `LessThan`, which can make it hard to determine why a point was rejected! One way to clarify this is by using a Function Alias- the name of your Function Alias will be printed.
        
*   Points which pass all Conditions but do not have the highest calculated Score are drawn as red cylinders, with their height proportional to their Score. They are labelled with their Score.
    

> *   The point which has the highest calculated Score is drawn as a green cylinder, and labelled with its Score

Reviewer 2: ```suggestion
*   The point which has the highest calculated Score is drawn as a green cylinder, and labelled with its Score.
```
